kart/spatial_filters.py

```diff
@@ -17,6 +17,44 @@
 # - handle the case where the spatial filter cannot or can only partially be projected to the target CRS
 
 
+def spatial_filter_options():
+    """
+    A decorator that can be added to a command to add "--spatial-filter" and "--spatial-filter-crs" options.
+    These can then be converted to a spatial filter using:
+    SpatialFilter.from_cli_opts(spatial_filter, spatial_filter_crs)
+    """
+
+    def decorator(f):
+        options = (
+            click.option(
+                "--spatial-filter",
```

Comment:
> --spatial-filter=<blob-sha> // clashes with (2)

AFAICT no valid hex WKB has a length of exactly 40 bytes, so a _full_ blob ID wouldn't be ambiguous.

Reply:
or 64 bytes for sha256?

Reply:
Polygon: 9 bytes + N*LinearRing
MultiPolygon: 9 bytes + N*Polygon
LinearRing: 4 bytes + N*Point
Point: 16 bytes

Now, a linear ring has to have at least 3 points, and a non-empty Polygon needs at least one ring, and a non-empty MultiPolygon needs at least one polygon, so seems like the smallest valid polygon is `9+4+16*3=61 bytes`, and the smallest valid multipolygon is `9+9+4+16*3=70 bytes`

I guess it works, seems a bit of a shady approach though 😉

Reply:
> so a full blob ID

ah, yeah, but partial blob ids: all bets are off? :-)

Reply:
I'd rather use the prefix than the length eg `0106000000010000000103000000010000000` - that seems to mean MULTIPOLYGON or something, I'd have to dig into it a bit more to see if it makes sense. If that approach makes sense then seems fine - otherwise maybe an explicit prefix like "hexwkb:0106000000010000000` but it's better if such things can be avoided and it Just Works - the user supplies what they imagine a valid input would be and we recognise what it is without them looking at the help pages.

Reply:
I guess my thinking is that a _user_ won't ever be providing hexwkb, in the same way that they won't provide anything but the very simplest WKT: it'd come from an app/something. So the _easy_ stuff has to be ref-name; WKT+EPSG code; file with CRS+WKT. And other stuff needs to be _possible_ but can be a bit more verbose.

Reply:
Just by the way, I've been using hexwkb, not wkt, to set spatial filters so far. Reason being: Kart diff outputs hexwkb, not wkt, so it's easier for me to get hexwkb (edit a working copy and then run kart diff).

There's far too many things we could support here - 
- The 5 or 7 or maybe 12  options above
- paths to git objects (as opposed to refs or SHAs of git objects)
- we could support both 2 parameter and 1 parameter (setting 2 parameters with 2 values is easier for users to delimit correctly, but we need to support a single-value syntax too for refs to work, so... we could do both)
- single-line or multi-line delimiter variations - ; is a good command line delimiter for single line usage, but inside a file, it makes sense to put things on separate lines

I propose a certain limited set of things to make it manageable - it's still quite a lot - most of them you recommend above:
1. Only one flag, `--spatial-filter`. Since this flag must sometimes contain both values (eg for `--spatial-filter=refname`, we avoid a bunch of weirdness if we say it always contains both values.
2. CRS must always be specified, there's no default.
3. CRS first, geometry second.
4. There are two basic syntaxes that `--spatial-filter` can have - either `--spatial-filter=CRS_SPEC;GEOMETRY_SPEC` or `--spatial-filter=COMBINED_SPEC`. So, semi-colon delimited.

5. The CRS_SPEC is either:
a) "EPSG:4167" or similar. 
b) An "@filename" which then contains either:
i) "EPSG:4167" or similar, or ii) a CRS definition in WKT.
Inlined CRS definition in WKT before the ; delimiter is not allowed. This is so the delimiter doesn't get lost in malformed WKT.

6. The GEOMETRY_SPEC is either:
a) WKT
b) hex-encoded WKB
c) An @filename which then contains a) or b)

7. The combined spec is either:
a) An @filename
b) A refname eg `refs/directory/file` or `[refs/filters/]filter-name`
c) A blob SHA

This must point to a file or blob which must contain all of the following:
i) CRS_SPEC - must be inlined, can be a short name eg "EPSG:4167" or similar, or a WKT definition
ii) A double line-feed (an empty line)
iii) GEOMETRY_SPEC - must be inlined, can be WKT or hex-encoded WKB

There's no real ambiguity in this system, except that technically a valid SHA and a valid refname could collide. This is a problem git already has a lot of - branch-names can collide with SHAs too - but in practise it doesn't happen except when users screw up. I guess we would just fail immediately if it's ambiguous what the user means (ie, the user wants a spatial filter of 9d5d8e7, where 9d5d8e7 is both a refname and a blob ID).


One open question - should you be notified if the refname you are started using as a spatial filter now points to a different spatial filter? ie "The spatial filter at refs/filters/xyz has changed - run `kart fetch --spatial-filter=refs/filters/xyz` to update your working copy with the new spatial filter." This would mean storing spatial filters by reference instead of by value

Reply:
> I propose a certain limited set of things to make it manageable - it's still quite a lot - most of them you recommend above:

Seems good. Could we drop 5B & 6C and instead solely use 7A for files?

> paths to git objects (as opposed to refs or SHAs of git objects)

Is slightly more complex to use in-repo file paths server-side, since when the user is using a filter for a _clone_ it's not resolvable. So I think what we pass with a _clone_ has to be either a resolved geometry+CRS or a ref/blob spec.

> I guess we would just fail immediately if it's ambiguous what the user means (ie, the user wants a spatial filter of 9d5d8e7, where 9d5d8e7 is both a refname and a blob ID).

Seems like just having a priority is fine? Resolve partial→full blob-sha; then full-blob-sha >> refname? Can always force to a ref with `/refs/...`

> This would mean storing spatial filters by reference instead of by value

I think we definitely want to store the refspec/blob-spec by reference. Gut feeling is filenames could/should also be stored by reference rather than value?

Q: Should `@filename` (probably don't need to ping whoever has that handle on Github 😉 ) be able to be absolute/relative/either? I guess because it's client-specified it can be any path we just resolve wrt the repository (working tree?) root.

Dunno how we deal with "contents of some random temp file" vs "save this file path"?

> One open question - should you be notified if the refname you are started using as a spatial filter now points to a different spatial filter?

Feels like we should fetch filters by default. And regardless of source — if the filter is updated then maybe we notify with `status` and update the WC whenever we change it next? Or can manually do with `checkout --`? Not quite sure  how the semantics/details of this would work, maybe worth examining how updating a sparse-checkout filter in Git works?

Reply:
> Seems good. Could we drop 5B & 6C and instead solely use 7A for files?

Yeah this. makes sense

> 
> > paths to git objects (as opposed to refs or SHAs of git objects)
> Is slightly more complex to use in-repo file paths server-side, since when the user is using a filter for a _clone_ it's not resolvable. So I think what we pass with a _clone_ has to be either a resolved geometry+CRS or a ref/blob spec.

Yeah let's definitely not do this then. It's not in my proposal anyway. One less thing.

> > I guess we would just fail immediately if it's ambiguous what the user means (ie, the user wants a spatial filter of 9d5d8e7, where 9d5d8e7 is both a refname and a blob ID).
> 
> Seems like just having a priority is fine? Resolve partial→full blob-sha; then full-blob-sha >> refname? Can always force to a ref with `/refs/...` (should store full blob-shas & full refs)
> 

> I think we definitely want to store the refspec/blob-spec by reference. Gut feeling is filenames could/should also be stored by reference rather than value?

I'm inclined to not store filenames by reference -
- as you say, the user could supply any random or temporary file from anywhere on their system - we don't know if they want us to consult that file indefinitely, or if they just downloaded into their ~/Downloads right now and they'll delete it if they ever decide ~/Downloads is too cluttered
- the repo config can now extends beyond the repo and potentially include a file in /tmp/ or ~/Downloads/ and the repo can break if either it or the target file is moved independently.
- if the user wants to update the spatial filter, they can easily edit the file and then run the appropriate command.

Saving refs or blob SHAs doesn't have this issue since it's very unlikely the user will accidentally break that - the blob will stay with the repo. And saving by value - inlining it in the config - is fine too.

> 
> Q: Should `@filename` (probably don't need to ping whoever has that handle on Github 😉 ) be able to be absolute/relative/either? I guess because it's client-specified it can be any path we just resolve wrt the repository (working tree?) root.

I think it's just resolved relative to the user's current working directory, which is fine as long as we are not saving it anywhere.

> 
> Dunno how we deal with "contents of some random temp file" vs "save this file path"?
> 
As above - not sure we should ever save the file path, safer just to save by value in this case.

> > One open question - should you be notified if the refname you are started using as a spatial filter now points to a different spatial filter?
> 
> Feels like we should fetch filters by default. And regardless of source — if the filter is updated then maybe we notify with `status` and update the WC whenever we change it next? Or can manually do with `checkout --`? Not quite sure how the semantics/details of this would work, maybe worth examining how updating a sparse-checkout filter in Git works?

I can try to update the spatial filter mostly-automatically the next time the WC is updated. The only obstacle I have thought of is if we try and do both of the following:
a) The user can change the spatial filter anytime, even if the WC is dirty
b) Having edits to features that are outside the spatial filter is not allowed (ie, they are assumed to be inserts that inadvertently reused earlier PK values, rather than anything the user is actually trying to do)

These two ideas seem to me to be incompatible and so if we do b), we should be aware that we can't do a), including that we can't automatically update the spatial filter at any time (ie, when running `kart fetch` and finding a new spatial filter, we may not immediately be able to apply it if the working copy is dirty).
But apart from this potential problem, it seems possible. Does that make sense?

Reply:
Mostly done, but the specify by ref / by objectid is turned off for now - I need to revisit it when I get spatially filtered clones to work, since it's actually karthub that needs to return an error to kart if you specify a filter that doesn't exist. So, punted that for now

Reply:
> Mostly done, but the specify by ref / by objectid is turned off for now - I need to revisit it when I get spatially filtered clones to work, since it's actually karthub that needs to return an error to kart if you specify a filter that doesn't exist. 

Should work client-side as well, not just for clones. But yes, can punt regardless.

> b) Having edits to features that are outside the spatial filter is not allowed

By "not allowed" do you mean "go away, you can't commit this"? I can't see this working well in reality, users will always add whatever they want :-)

Reply:
Agreed, users do whatever they want. But there is a bad scenario:

- Layer X has feature PKs from 1 to 1 million.
- User spatially filters layer X, WC only gets features with PKs 1000 to 2000.
- User inserts a new feature in WC with PK 2001.

The ambiguity here is what did the user intend to do - did they want to edit pre-existing feature 2001, or, were they trying to insert a new feature and they just accidentally reused an existing PK that they couldn't see? We can probably mostly prevent this from happening by messing with the sequence generators in the working copies, so that if the user inserts a new feature, they'll get PK=one million and one, but as you say, users do whatever they want, so it can still happen that the user has inserted a feature with PK 2001, which is the PK of a feature that already exists but that we didn't write to their working copy. Let's call this a "possibly-accidental-PK-collision" When this happens we have to do one of the following:

1. Just accept their blind-ish edit to feature with PK=2001 even though they may have intended to insert a new feature. (Blind-ish because they didn't make the edit by editing the existing feature, but not completely blind since they can at least see what edit does by running `kart diff` afterwards)
2. Accept their insert - commit the insert as they inserted it, except with a fresh PK of one million and one.
3. Prompt / warn / fail somehow.

Now, all of those except for 1) depend on the current spatial filter. For 2 and 3, if the spatial filter changes, it means the semantics of dirty rows in the working copy changes too - for instance, if we choose 3 then when the user tries to commit we are simply going to warn or fail with an error message because the user has inserted a new feature that reuses PK=2001, a PK outside their spatial filter. But if instead of committing, the user changes their spatial filter, then this would change which dirty rows in their working copy are now edits we can accept, and which are disallowed as probably-accidental-PK-collisions that we have to warn or fail for. Approach 2 has similar problems - we can half-solve the problem by going through the working copy and assigning new PKs to "possible-accidental-PK-collision" features before we change the spatial filter, but we can't fix the opposite problem where things that were originally perfectly reasonable edits will now be outside the spatial filter and will now look like possible-accidental-PK-collisions.

Anyway, so we don't have to do approach 2 or 3, but if we don't, the user might accidentally overwrite old features that they weren't intending to. That's what possible-accidental-PK-collision means, after all. But if we DO implement approach 2 or 3, we have to be aware that it means that we can't just change the spatial filter automatically whenever we want (ie, if they pull a new version of it from the origin) - we might first have to get the user to make the working copy clean by committing or discarding changes (or at least clean of problem changes). Similarly, if the user manually tries to change their spatial filter, we would maybe have to ask them to commit or reset some edits first.

Reply:
Yes, PK clashes we've got to try and help prevent, I think it's too much of a footgun otherwise. I guess I'm trying to differentiate here between "all editing must be inside this box" and "you can't add a new feature that with a PK that exists outside the filter". Seems like the former we don't need to prevent, you're primarily talking about the latter issue?

Maybe we treat the spatial filter a bit like how submodules in git work ux-wise? They don't auto-update, but they're presented in various commands as being out-of-sync.

I wonder if the config shows the desired filter (ref/sha/geometry/etc), and the _current_ filter is written elsewhere so we can easily compare them and determine whether they're in sync and needs to be explicitly re-applied? Then the explicit re-applying can potentially require a clean WC or deal with clashes.

Reply:
Unpunted the refs / object IDs variant - it now works when doing a clone. Added tests.
(It doesn't work during `kart init` since there's no refs or object IDs - and it doesn't work during kart checkout, since I haven't yet implemented the ability to change spatial filters after the fact).

The only filter that suddenly can be out of date due to apparently unrelated command is the ref one - none of the other options can change value except by the user explicitly changing them, so we don't need to be able to detect if they are out of date. For the ref one I'm currently storing both the ref and the target it resolves to, so we can detect when the ref's target has changed.

We may also want to store a hash of spatial filter in the working copy, just like we store the tree that the WC is synced to, to make sure the WC stays in sync with the config (then we could detect even if the user had manually edited the config file).